### PR TITLE
refactor(remote ext): use jsonrpsee 

### PR DESCRIPTION
**Pull request by niklasad1**
Fixes https://github.com/paritytech/substrate/pull/8038#discussion_r578991856 which replaces `jsonrpc-core-client` with `jsonrpsee-http-client` and doesn't require `tokio 0.1`

### CODE REVIEW

File: utils/frame/try-runtime/remote-externalities/Cargo.toml [hunk 1]
```diff
@@ -13,16 +13,13 @@ readme = "README.md"
 targets = ["x86_64-unknown-linux-gnu"]
 
 [dependencies]
-jsonrpc-core-client = { version = "15.1.0", features = ["http"] }
-sc-rpc-api = { version = "0.9.0", path = "../../../../client/rpc-api" }
-sc-rpc = { version = "3.0.0", path = "../../../../client/rpc" }
-futures = "0.1.29"
+jsonrpsee-http-client = { git = "https://github.com/paritytech/jsonrpsee", default-features = false, features = ["hyper-client"] }
```
Comment (niklasad1) [inline utils/frame/try-runtime/remote-externalities/Cargo.toml:16]: this works with `tokio 0.2` and `async std`

Comment (kianenigma) [reply utils/frame/try-runtime/remote-externalities/Cargo.toml:16]: can you pin to a rev?

Comment (kianenigma) [reply utils/frame/try-runtime/remote-externalities/Cargo.toml:16]: also, @gnunicorn I recall that you had some grudge against using git deps, and that we had to get rid of them prior to releases? Is this okay, given that the code is only used in a niche testing cli utility.

File: utils/frame/try-runtime/remote-externalities/src/lib.rs [hunk 2]
```diff
@@ -261,43 +269,24 @@ impl Builder {
...
-		let mut rt = tokio::runtime::Runtime::new().expect("Unable to create a runtime");
-		let uri = self.as_online().uri.clone();
-		rt.block_on::<_, _, ()>(futures::lazy(move || {
-			trace!(target: LOG_TARGET, "rpc: finalized_head");
-			let client: sc_rpc_api::chain::ChainClient<Number, Hash, (), ()> =
-				jsonrpc_core_client::transports::http::connect(&uri).wait().unwrap();
-			Ok(client.finalized_head().wait().unwrap())
-		}))
-		.unwrap()
+		// TODO: move client to the builder.
```
Comment (kianenigma) [inline utils/frame/try-runtime/remote-externalities/src/lib.rs:272]:
```suggestion
```

File: utils/frame/try-runtime/remote-externalities/Cargo.toml (hunk 1 above)
Comment (gnunicorn) [reply utils/frame/try-runtime/remote-externalities/Cargo.toml:16]: the problem is that crates.io doesn't allow for git-dependencies and as this is _not_ marked `publish = false`, this means this crate will fail when we try to publish.

Comment (niklasad1) [reply utils/frame/try-runtime/remote-externalities/Cargo.toml:16]: right, I think we could release alpha/beta release for the JSONRPC clients only.

//cc @dvdplm @maciejhirsz

Comment (kianenigma) [reply utils/frame/try-runtime/remote-externalities/Cargo.toml:16]: yeah, let's aim for having a release of jsonrpsee as well while we're at it.

**Comment by kianenigma** [comment]
Waht's the prospect of this?

**Comment by niklasad1** [comment]
> Waht's the prospect of this?

After https://github.com/paritytech/jsonrpsee/pull/215, I plan to publish an alpha release for the clients

File: utils/frame/remote-externalities/src/lib.rs [hunk 3]
```diff
@@ -250,18 +259,8 @@ impl Builder {
...
-			jsonrpc_core_client::transports::http::connect(&uri)
-				.compat()
-				.map_err(|_| "client initialization failed")
-				.await?;
-		client
-			.storage_pairs(prefix, Some(at))
-			.compat()
-			.map_err(|_| "rpc finalized_head failed.")
-			.await
+		RpcApi::storage_pairs(&*self.as_online().rpc, prefix, at).await.map_err(|_| "rpc finalized_head failed.")
```
Comment (kianenigma) [inline utils/frame/remote-externalities/src/lib.rs:263]:
```suggestion
		RpcApi::storage_pairs(&*self.as_online().rpc, prefix, at).await.map_err(|_| "rpc state_getPairs failed.")
```

File: utils/frame/remote-externalities/src/lib.rs [hunk 4]
```diff
@@ -101,9 +101,13 @@
 //! }
 //! ```
 
+// jsonrpsee_proc_macros generates faulty warnings: https://github.com/paritytech/jsonrpsee/issues/106
```
Comment (kianenigma) [inline utils/frame/remote-externalities/src/lib.rs:104]: Would be good if you provide follow ups for this.

**Review by kianenigma — APPROVED** [review]

File: utils/frame/remote-externalities/Cargo.toml [hunk 5]
```diff
@@ -13,16 +13,15 @@ readme = "README.md"
 targets = ["x86_64-unknown-linux-gnu"]
 
 [dependencies]
-jsonrpc-core-client = { version = "15.1.0", features = ["http"] }
-sc-rpc-api = { version = "0.9.0", path = "../../../client/rpc-api" }
-sc-rpc = { version = "3.0.0", path = "../../../client/rpc" }
-futures = "0.3"
+jsonrpsee-http-client = { version = "0.2.0-alpha", default-features = false, features = ["tokio02"] }
+# Needed by `jsonrpsee-proc-macros`
```
Comment (bkchr) [inline utils/frame/remote-externalities/Cargo.toml:17]: I hope there exists an upstream issue to fix this. Having macros depend on other crates isn't such a nice experience.

**Review by bkchr — APPROVED** [review]

File: utils/frame/remote-externalities/src/lib.rs [hunk 6]
```diff
@@ -161,7 +175,13 @@ pub struct OnlineConfig {
 impl Default for OnlineConfig {
 	fn default() -> Self {
 		Self {
-			uri: "http://localhost:9933".into(),
+			uri: TARGET.to_owned(),
+			rpc: Arc::new(
+				HttpClient::new(
+					TARGET,
```
Comment (kianenigma) [inline utils/frame/remote-externalities/src/lib.rs:181]: okay I approved but this is very wrong :D you must usethe given uri, not the hardcoded stuff. I will have to fix this a bit.

File: utils/frame/remote-externalities/Cargo.toml (hunk 5 above)
Comment (niklasad1) [reply utils/frame/remote-externalities/Cargo.toml:17]: https://github.com/paritytech/jsonrpsee/issues/214

File: utils/frame/remote-externalities/src/lib.rs (hunk 4 above)
Comment (niklasad1) [reply utils/frame/remote-externalities/src/lib.rs:104]: it's fixed on master but I haven't done another release yet :)

File: utils/frame/remote-externalities/Cargo.toml (hunk 5 above)
Comment (niklasad1) [inline utils/frame/remote-externalities/Cargo.toml:17]:
```suggestion
# Needed by jsonrpsee-proc-macros: https://github.com/paritytech/jsonrpsee/issues/214
```

File: utils/frame/remote-externalities/src/lib.rs [hunk 7]
```diff
@@ -161,7 +175,13 @@ pub struct OnlineConfig {
 impl Default for OnlineConfig {
 	fn default() -> Self {
 		Self {
-			uri: "http://localhost:9933".into(),
+			uri: TARGET.to_owned(),
+			rpc: Arc::new(
+				HttpClient::new(
```
Comment (kianenigma) [inline utils/frame/remote-externalities/src/lib.rs:180]: This is wrong -- We can't initialize the client here. Then, if we override the `at` later, if will be a noop. Will push fixes.

Comment (kianenigma) [reply utils/frame/remote-externalities/src/lib.rs:180]: (or maybe it is correct -- have to check)

**Comment by kianenigma** [comment]
bot merge

File: utils/frame/remote-externalities/src/lib.rs (hunk 7 above)
Comment (niklasad1) [reply utils/frame/remote-externalities/src/lib.rs]: ok, I didn't understand that the URL/URI could be replaced thanks for the fix.

**Comment by niklasad1** [comment]
bot merge

**Pull request merged**